Export meshes as OBJ

andburn opened the pull request:
Implemented basic OBJ exporting, works for the unity 4 & 5 bundles I have.

Elements not implemented:
- *topology* triangles
- *compressed meshes*
- *half-floats* (16 bits)

Don't know how widespread these missing features are, but I didn't encounter them.

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+	def export(self):
+		vns = self.vertices["normals"]
+		vts = self.vertices["uv1"]
+		out_str = ""
```
This is very slow. Create a list `ret = []`, and do `return "\n".join(ret)`.

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+					out += str(i)
+				out += "/"
+				if vn:
+					out += str(i)
```
Im not sure this works correctly if only one of vt or vn is non-null.

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+	def face_str(self, indices, vts, vns):
+		out = "f "
+		# reverse indices
+		inds = indices[::-1]
```
You don't need a temporary variable, just do `for i in indices[::-1]`

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+		vt = vts > 0
+		vn = vns > 0
+		for idx in inds:
+			i = idx + 1 # OBJ indices 1 based
```
Same here, just do`i + 1`

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+class OBJMesh:
+	def __init__(self, mesh):
+		self.mesh = mesh
+		if (self.mesh.mesh_compression):
```
extraneous parentheses

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+	streams = []
+	# scan the channel's stream value for distinct entries
+	for c in channels:
+		if c['stream'] not in streams:
```
Bad quoting ;)

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,223 @@
+		self.__alt_rgba()
+		return self
+
+	def __alt_rgba(self):
```
what's this for?

andburn commented:
Updated:
- change string concatenation to list joins
- remove single quotes
- remove `Vec4` rgba aliases (not used)
- remove unnecessary temporary variables
- fix possible trailing slash on face definitions, if normals were empty

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+	return len(streams)
+
+
+def extract_indicies(mesh):
```
`indices`

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+		indices.append(sub_indices)
+		triangles.append(sub_triangles)
+
+	return {
```
I don't think you should be creating a dict for the return here. you can just `return indices, triangles`

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+		if self.mesh.mesh_compression:
+			# TODO handle compressed meshes
+			raise NotImplementedError("(%s) compressed meshes are not supported" % (mesh.name))
+		else:
```
extraneous else clause

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+
+		for i in range(0, sub_count):
+			if sub_count == 1:
+				ret.extend(["usemtl ", self.mesh.name, "\n"])
```
You should be doing `ret.append("usemtl %s\n" % (self.mesh.name))`

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+			if sub_count == 1:
+				ret.extend(["usemtl ", self.mesh.name, "\n"])
+			else:
+				ret.extend(["usemtl ", self.mesh.name, "_", str(i), "\n"])
```
Same here.

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+	def vec3_str(self, prefix, vec):
+		return "%s %s %s %s\n" % (prefix, -vec.x, vec.y, vec.z)
+
+	def face_str(self, indices, vts_len, vns_len):
```
All the `_str` functions can be staticmethods.

jleclanche replied:
Come to think of it, you can just implement `vecX_str` on the class itself, which will let you do `str(vec)`. Much simpler.

jleclanche commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+		for v in vts:
+			ret.append(self.vec2_str("vt", v))
+
+		ret.extend(["\ng ", self.mesh.name, "\ns 1\n" ])
```
This is not very readable.

andburn commented on unitypack/export.py:
```diff
@@ -0,0 +1,212 @@
+		# TODO investigate possible alignment here, after each stream
+
+	return {
+		"vertices": vertices,
```
Are you ok with using the dictionary here, or is there a more pythonic way of doing it?

jleclanche replied:
Well, pythonically there's two options:
- Create a struct-like class with those attributes (or even just a namedtuple)
- Return multiple values... but with the amount of values being returned here, it gets unreadable.

andburn commented:
Updated:
- Comments addressed.
- Make `vecX_str` methods static (I think they fit better inside the `OBJMesh` class, as they are specific to the OBJ format).
- Added `MeshData` class to hold vertex and index info, moved top-level methods into it too.

jleclanche commented:
> I think they fit better inside the OBJMesh class, as they are specific to the OBJ format

I wouldn't disagree if these were generic vectors but... just call them OBJVector2, OBJVector3 etc and you should be fine.

andburn commented:
Ok.
One other thing. Its outputting the values to 15 decimal places, does rounding it down to 6-8 seem sensible?

jleclanche commented:
We can figure that out another time.

andburn commented:
Updated to use `__str__`

jleclanche commented:
Thank you!
